commands/dht.go

```diff
@@ -0,0 +1,163 @@
+package commands
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	cmds "gx/ipfs/QmQtQrtNioesAWtrx8csBvfY37gTe94d6wQ3VikZUjxD39/go-ipfs-cmds"
+	"gx/ipfs/QmR8BauakNcBa3RbE4nbQu76PDiJgoQgz8AJdhJuiU4TAw/go-cid"
+	pstore "gx/ipfs/QmRhFARzTHcFh8wUxwN5KvyTGq73FLC65EfFAhz8Ng7aGb/go-libp2p-peerstore"
+	notif "gx/ipfs/QmWaDSNoSdSXU9b6udyaq9T8y6LkzMwqWxECznFqvtcTsk/go-libp2p-routing/notifications"
+	cmdkit "gx/ipfs/Qmde5VP1qUkyQXKCfmEUA7bP64V2HAptbJ7phuPp7jXWwg/go-ipfs-cmdkit"
+)
+
+const (
+	dhtVerboseOptionName   = "verbose"
+	numProvidersOptionName = "num-providers"
+)
+
+// Note, most of this is copied directly from go-ipfs (https://github.com/ipfs/go-ipfs/blob/master/core/commands/dht.go).
+// A few simple modifications have been adapted for filecoin.
+var dhtCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline:          "Issue commands directly through the DHT.",
+		ShortDescription: ``,
+	},
+
+	Subcommands: map[string]*cmds.Command{
+		"findprovs": findProvidersDhtCmd,
+	},
+}
+
+var findProvidersDhtCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline:          "Find peers that can provide a specific value, given a key.",
+		ShortDescription: "Outputs a list of newline-delimited provider Peer IDs.",
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("key", true, false, "The key to send provide records for.").EnableStdin(),
+	},
+	Options: []cmdkit.Option{
+		cmdkit.BoolOption(dhtVerboseOptionName, "v", "Print extra information."),
+		cmdkit.IntOption(numProvidersOptionName, "n", "The number of providers to find.").WithDefault(20),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		numProviders, _ := req.Options[numProvidersOptionName].(int)
+		if numProviders < 1 {
+			return fmt.Errorf("number of providers must be greater than 0")
+		}
+
+		c, err := cid.Parse(req.Arguments[0])
+		if err != nil {
+			return err
+		}
+
+		ctx, cancel := context.WithCancel(req.Context)
+		ctx, events := notif.RegisterForQueryEvents(ctx)
+
+		pchan := GetPorcelainAPI(env).RouterFindProvidersAsync(ctx, c, numProviders)
+
+		go func() {
+			defer cancel()
+			for p := range pchan {
+				np := p
+				notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+					Type:      notif.Provider,
+					Responses: []*pstore.PeerInfo{&np},
+				})
+			}
+		}()
+		for e := range events {
+			if err := res.Emit(e); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *notif.QueryEvent) error {
+			pfm := pfuncMap{
+				notif.FinalPeer: func(obj *notif.QueryEvent, out io.Writer, verbose bool) {
+					if verbose {
+						fmt.Fprintf(out, "* closest peer %s\n", obj.ID) // nolint: errcheck
```

Comment:
The ipfs encoders don't do error checking when printing to the writer.  I put a bandaid on this with the "nolint" comments.  Happy to think through error checking if this seems important.